xdsl/dialects/onnx.py

```diff
@@ -860,6 +860,79 @@ def __init__(self, func: Attribute):
         )
 
 
+@irdl_op_definition
+class MatMul(IRDLOperation):
+    """
+    The operation MatMul performs matrix multiplication between two input matrices, A and B, and returns the result as matrix Y.
+    Matrix multiplication is a fundamental operation in linear algebra, where each element of the resulting matrix Y is computed by taking the
+    dot product of the corresponding row of matrix A and column of matrix B.
+    """
+
+    name = "onnx.MatMul"
+
+    # describe annotated type
+    T = Annotated[AnyFloat | IntegerType, ConstraintVar("T")]
+
+    # input matrices
+    matrix_A = operand_def(TensorType[T])
+    matrix_B = operand_def(TensorType[T])
+
+    # output matrices
+    matrix_Y = result_def(TensorType[T])
+
+    assembly_format = (
+        "`(` $matrix_A `,` $matrix_B `)` attr-dict `:` `(` type($matrix_A) `,"
+        "` type($matrix_B) `)` `->` type($matrix_Y) "
+    )
+
+    def __init__(
+        self,
+        matrix_A: SSAValue,
+        matrix_B: SSAValue,
+    ):
+        super().__init__(
+            operands=[matrix_A, matrix_B],
+            result_types=[matrix_A.type],
+        )
+
+    def verify_(self) -> None:
+        # store dimensions of tensor A and tensor B
+        res_shape: list[int] = []
+        if (
+            not isinstance(matrix_A_type := self.matrix_A.type, TensorType)
+            or not isinstance(matrix_B_type := self.matrix_B.type, TensorType)
+            or not isinstance(matrix_Y_type := self.matrix_Y.type, TensorType)
+        ):
+            raise VerifyException("input matrix A should be a 2D tensor")
```

Comment:
these are checked by the irdl generated verifier, and can be casts instead
```suggestion
        matrix_A_type = cast(TensorType[Any], self.matrix_A.type)
        matrix_B_type = cast(TensorType[Any], self.matrix_B.type)
        matrix_Y_type = cast(TensorType[Any], self.matrix_Y.type)
```

Reply:
Done